### PR TITLE
Logaddexp for complex in CPU

### PR DESCRIPTION
**Pull request by mfkasim1**
Continuation of PR #93153 where I implemented logaddexp for complex, but didn't expose it to `torch.logaddexp`. So this PR is to expose the complex logaddexp to `torch.logaddexp`.

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

### CODE REVIEW

File: aten/src/ATen/native/cpu/ReduceOpsKernel.cpp [hunk 1]
```diff
@@ -113,57 +114,6 @@ static void cumprod_cpu_kernel(const Tensor& result, const Tensor& self, int64_t
     );
   });
 }
-// custom min and max to be used in logcumsumexp for complex arguments
```
Comment (lezcano) [inline aten/src/ATen/native/cpu/ReduceOpsKernel.cpp:116]: Actually, why are we shuffling things around? Could you revert this change? Often things are structured certain way to improve compilation times (C++ is best compiled in large files, while CUDA is best compiled in smaller files).

File: torch/_refs/__init__.py [hunk 2]
```diff
@@ -1447,11 +1447,28 @@ def le(a: TensorLikeType, b: TensorLikeType) -> TensorLikeType:
...
     # Nb. this implementation does nto distribute the gradients evenly when a == b
-    mask = a >= b
+    mask = torch.real(a) >= torch.real(b)
     max_ = torch.where(mask, a, b)
     min_ = torch.where(mask, b, a)
-    inf_mask = torch.logical_and(torch.isinf(a), a == b)
-    return torch.where(inf_mask, a, max_ + torch.log1p(torch.exp(min_ - max_)))
+    inf_mask = torch.logical_and(
+        torch.isinf(torch.real(a)), torch.real(a) == torch.real(b)
+    )
```
Comment (lezcano) [inline torch/_refs/__init__.py:1455]: This does not follow core, as there it uses `std::isfinite` which is equivalent to `torch.isfinite`, which is not equivalent to `not torch.isinf`.

**Review by lezcano — COMMENTED** [review]
Try keeping diffs small when possible, please :)

File: aten/src/ATen/native/cpu/ReduceOpsKernel.cpp (hunk 1 above)
Comment (mfkasim1) [reply aten/src/ATen/native/cpu/ReduceOpsKernel.cpp:116]: I'm moving this part of the code because it's going to be used in 2 cpp places: in ReduceOpsKernel.cpp and in BinaryOpsKernel.cpp. If I reverted this change, I'm not sure if I can load the function defined in ReduceOpsKernel.cpp to be used in BinaryOpsKernel.cpp.

File: torch/_refs/__init__.py (hunk 2 above)
Comment (mfkasim1) [reply torch/_refs/__init__.py]: Done in the latest commit. Changed it to `torch.logical_not(torch.isfinite(...))`

File: aten/src/ATen/native/cpu/LogAddExp.h [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+c10::complex<scalar_t> _log_add_exp_helper(const c10::complex<scalar_t>& x, const c10::complex<scalar_t>& y) {
+  auto min = _logcumsumexp_minmax<scalar_t, true>(x, y);
+  auto max = _logcumsumexp_minmax<scalar_t, false>(x, y);
+  auto min_real = std::real(min);
+  auto max_real = std::real(max);
+
+  if (at::_isnan(min)) {  // either real is nan or imag is nan
+    // handling the "infectious" NaNs
+    return {std::numeric_limits<scalar_t>::quiet_NaN(), std::numeric_limits<scalar_t>::quiet_NaN()};
+  } else if ((!std::isfinite(min_real)) && (min_real == max_real)) {
```
[the remark this answers is not in the document]
Comment (mfkasim1) [reply aten/src/ATen/native/cpu/LogAddExp.h]: Done

File: torch/_refs/__init__.py [hunk 4]
```diff
@@ -1447,11 +1447,28 @@ def le(a: TensorLikeType, b: TensorLikeType) -> TensorLikeType:
...
+        inf_vals = torch.where(
+            neg_min_mask, min_, torch.log(torch.exp(min_) + torch.exp(max_))
+        )
+        non_nan_vals = torch.where(
+            inf_mask, inf_vals, max_ + torch.log1p(torch.exp(min_ - max_))
+        )
+        # the type for full_like does not include tensor yet
+        nan_vals = torch.full_like(complex_tens, complex(float("nan"), float("nan")))  # type: ignore[arg-type]
+        nan_mask = torch.isnan(min_)
+        return torch.where(nan_mask, nan_vals, non_nan_vals)
```
Comment (lezcano) [inline torch/_refs/__init__.py:1469]: No need to instantiate the full tensor, as broadcasting works.
```suggestion
        nan_mask = torch.isnan(min_)
        return torch.where(nan_mask, nan_vals, complex(float("nan"), float("nan")))
```

File: aten/src/ATen/native/cpu/BinaryOpsKernel.cpp [hunk 5]
```diff
@@ -893,6 +894,14 @@ void logaddexp_kernel(TensorIteratorBase& iter) {
               (a1 == b1) & (a1.abs() == inf));
           return convert_float_bfloat16(a0, a1);
         });
+  } else if (isComplexType(iter.dtype())) {
+    AT_DISPATCH_COMPLEX_TYPES(iter.dtype(), "logaddexp_cpu", [&]() {
+      cpu_kernel(
+        iter,
+        [=](scalar_t a, scalar_t b) -> scalar_t {
+          return _log_add_exp_helper(a, b);
+        });
```
Comment (lezcano) [inline aten/src/ATen/native/cpu/BinaryOpsKernel.cpp:903]:
```suggestion
      cpu_kernel(iter, _log_add_exp_helper<scalar_t>);
```

**Review by lezcano — COMMENTED** [review]
I believe the new `minmax` is not quite equivalent to the previous one. Once we fix that this should be ready to go.

File: aten/src/ATen/native/cpu/LogAddExp.h [hunk 6]
```diff
@@ -7,14 +7,14 @@ namespace at { namespace native {
...
-template <typename scalar_t, bool min>
-c10::complex<scalar_t> _logcumsumexp_minmax(c10::complex<scalar_t> x, c10::complex<scalar_t> y) {
+template <typename scalar_t>
+std::pair<c10::complex<scalar_t>, c10::complex<scalar_t>> _logcumsumexp_minmax(c10::complex<scalar_t> x, c10::complex<scalar_t> y) {
   if (at::_isnan(y)) {  // either real is nan or imag is nan
-    return y;
+    return std::make_pair(y, y);
   } else if (at::_isnan(x)) {  // either real is nan or imag is nan
-    return x;
+    return std::make_pair(x, x);
```
[the remark this answers is not in the document]
Comment (mfkasim1) [reply aten/src/ATen/native/cpu/LogAddExp.h:15]: Previous code calls this function like below:
```
auto min = _logcumsumexp_minmax<scalar_t, true>(x, y);
auto max = _logcumsumexp_minmax<scalar_t, false>(x, y);
```
So if `y` has nans (regardless whether `x` has nans or not), min and max both are `y`. If `y` does not have nans, but `x` does, min and max both are `x`. Am I missing something here?

[the remark this answers is not in the document]
Comment (lezcano) [reply aten/src/ATen/native/cpu/LogAddExp.h]: Nope, you are right! I was thinking that perhaps this function is not as equivalent to its real counterpart as we would like (it returns (y, y) by default rather than, say, (y,x) or smth), but the thing is that, if one of the inputs is NaN, then we'll return a complex(NaN,NaN), so it doesn't really matter.

**Review by lezcano — APPROVED** [review]
Cool! Thank you for the contribution!

**Comment by mfkasim1** [comment]
Thanks for your help and your reviews @lezcano!
@pytorchbot merge

File: aten/src/ATen/native/cpu/BinaryOpsKernel.cpp (hunk 5 above)
Comment (lezcano) [reply aten/src/ATen/native/cpu/BinaryOpsKernel.cpp:903]: yeah, I was not 100% sure about this one. Templates are difficult.

File: torch/_refs/__init__.py [hunk 7]
```diff
@@ -1447,11 +1447,26 @@ def le(a: TensorLikeType, b: TensorLikeType) -> TensorLikeType:
 )
 def logaddexp(a: TensorLikeType, b: TensorLikeType) -> TensorLikeType:
     # Nb. this implementation does nto distribute the gradients evenly when a == b
```
Comment (Skylion007) [inline torch/_refs/__init__.py:1449]: mind fixing this typo (nto) too while touching this file? :P

Comment (mfkasim1) [reply torch/_refs/__init__.py]: Sure, no prob
